### PR TITLE
Updated the soldeer version to 0.2.18 and added extra CLI tests

### PR DESCRIPTION
**Pull request by mario-eth**
## Motivation

We added new functionality to Soldeer, using a git repo as a dependency.
https://github.com/mario-eth/soldeer/issues/48

This feature has been requested even tho I do not encourage abusing it but it can be used for various unpublished repos.

## Solution
Now we can use git to pull dependencies. It supports github and gitlab

Example:
```bash
forge soldeer install test-project~v1 git@github.com:test/test.git
forge soldeer install test-project~v1 git@gitlab.com:test/test.git
```

```bash
forge soldeer install test-project~v1 https://github.com/test/test.git
forge soldeer install test-project~v1 https://gitlab.com/test/test.git
```

Or using custom commit hashes
```bash
forge soldeer install test-project~v1 git@github.com:test/test.git 345e611cd84bfb4e62c583fa1886c1928bc1a464
```

### CODE REVIEW

**Comment by DaniPopes** [comment]
Hey, this adds 3 C libraries as dependencies as you can see by the Cargo.lock diff. I would really prefer if this was not the case.

We dropped the git2 library a while ago in favor of just using the git CLI, see https://github.com/foundry-rs/foundry/blob/0e07ca51a2947f86d099645fb6eb6d0e47a4410d/crates/cli/src/utils/mod.rs#L276

In your case you can probably just get away with running `Command::new("git").args(["clone", a, b])`

**Review by DaniPopes — CHANGES_REQUESTED** [review]
Hey, this adds 3 C libraries as dependencies as you can see by the Cargo.lock diff. I would really prefer if this was not the case.

We dropped the git2 library a while ago in favor of just using the git CLI, see https://github.com/foundry-rs/foundry/blob/0e07ca51a2947f86d099645fb6eb6d0e47a4410d/crates/cli/src/utils/mod.rs#L276

In your case you can probably just get away with running `Command::new("git").args(["clone", a, b])`

File: crates/config/src/soldeer.rs [hunk 1]
```diff
@@ -13,6 +13,10 @@ pub struct MapDependency {
     /// The url from where the dependency was retrieved
     #[serde(default, skip_serializing_if = "Option::is_none")]
     pub url: Option<String>,
+
+    /// The commit in case git is used as dependency retrieval
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub commit: Option<String>,
```
Comment (DaniPopes) [inline crates/config/src/soldeer.rs:19]: this can be `rev` like in cargo, which can be a commit or a tag

File: crates/forge/tests/cli/soldeer.rs [hunk 2]
```diff
@@ -51,6 +51,95 @@ forge-std = "1.8.1"
...
+libs = ["lib"]
+
+# See more config options https://github.com/foundry-rs/foundry/blob/master/crates/config/README.md#all-options
+
+[dependencies]
+forge-std = { version = "1.8.1", url = "git@gitlab.com:mario4582928/Mario.git", commit = "bedb1775bd60e3e142a6db863982254a4b891b20" }
+"#;
+
+    let actual_foundry_contents = read_file_to_string(&foundry_file);
+    assert_eq!(foundry_contents, actual_foundry_contents);
```
Comment (DaniPopes) [inline crates/forge/tests/cli/soldeer.rs:140]: We've recently updated our test suite, you should use `assert_data_eq!(read_file_to_string(x), str![[""]])` and then run the test to update the assertion automatically.

**Comment by mario-eth** [comment]
> Hey, this adds 3 C libraries as dependencies as you can see by the Cargo.lock diff. I would really prefer if this was not the case.
> 
> 
> 
> We dropped the git2 library a while ago in favor of just using the git CLI, see https://github.com/foundry-rs/foundry/blob/0e07ca51a2947f86d099645fb6eb6d0e47a4410d/crates/cli/src/utils/mod.rs#L276
> 
> 
> 
> In your case you can probably just get away with running `Command::new("git").args(["clone", a, b])`

Thanks for the suggestion. I am gonna try to get around just with the cli then.
There are a few commands i need, not only clone but nothing too complicated.
Thanks

File: crates/forge/bin/cmd/soldeer.rs [hunk 3]
```diff
@@ -6,6 +6,8 @@ use soldeer::commands::Subcommands;
 // CLI arguments for `forge soldeer`.
 #[derive(Clone, Debug, Parser)]
 #[clap(override_usage = "forge soldeer install [DEPENDENCY]~[VERSION] <REMOTE_URL>
+    forge soldeer install [DEPENDENCY]~[VERSION] <GIT_URL>
+    forge soldeer install [DEPENDENCY]~[VERSION] <GIT_URL> <COMMIT>
```
Comment (DaniPopes) [inline crates/forge/bin/cmd/soldeer.rs:10]: same here, the CLI options should probably mimic the config keys, so `--rev`

**Review by DaniPopes — CHANGES_REQUESTED** [review]

**Comment by mario-eth** [comment]
@DaniPopes 
Solved all the comments:
https://github.com/mario-eth/soldeer/pull/98

**Review by DaniPopes — APPROVED** [review]

**Review by mattsse — APPROVED** [review]

**Pull request merged**